src/opentelemetry_instrumentation_kstreams/utils.py

```diff
@@ -101,57 +117,101 @@ def extract_consumer_client_id(instance: Stream) -> str:
         return instance.consumer._client._client_id
 
 
-def _enrich_span(
+def _enrich_base_span(
     span: Span,
     bootstrap_servers: List[str],
     topic: str,
-    partition: Optional[int],
     client_id: Optional[str],
-):
+) -> None:
+    """
+    Enriches the given span with Kafka-related attributes.
+
+    Used to enrich consumer and producer spans.
+
+    Args:
+        span: The span to enrich.
+        bootstrap_servers: List of Kafka bootstrap servers.
+        topic: The Kafka topic name.
+        partition: The partition number, if available.
+        client_id: The Kafka client ID, if available.
+        offset: The message offset, if available. Defaults to None.
+    Returns:
+        None
+    """
+    if not span.is_recording():
+        return
+
     span.set_attribute(SpanAttributes.MESSAGING_SYSTEM, "kafka")
     span.set_attribute(SpanAttributes.MESSAGING_DESTINATION, topic)
     span.set_attribute(SpanAttributes.MESSAGING_URL, json.dumps(bootstrap_servers))
     span.set_attribute(SpanAttributes.MESSAGING_DESTINATION_KIND, "topic")
+
     if client_id is not None:
         span.set_attribute(SpanAttributes.MESSAGING_KAFKA_CLIENT_ID, client_id)
 
-    if span.is_recording():
-        if partition is not None:
-            span.set_attribute(SpanAttributes.MESSAGING_KAFKA_PARTITION, partition)
+
+def _enrich_span_with_record_info(
+    span: Span, topic: str, partition: Optional[int], offset: Optional[int] = None
+) -> None:
+    """Used both by consumer and producer spans
+
+    It's in a different function because send needs to be injected with the span
+    info, but we want to be able to keep updating the span with the record info
+
+    Args:
+        span: The span to enrich.
+        topic: The Kafka topic name.
+        partition: The partition number, if available.
+        offset: The message offset, if available. Defaults to None.
+    Returns:
+        None
+    """
+    if offset is not None:
+        span.set_attribute(SpanAttributes.MESSAGING_KAFKA_MESSAGE_OFFSET, offset)
+    if partition is not None:
+        span.set_attribute(SpanAttributes.MESSAGING_KAFKA_PARTITION, partition)
+    if offset is not None and partition is not None:
+        span.set_attribute(
+            SpanAttributes.MESSAGING_MESSAGE_ID,
+            f"{topic}.{partition}.{offset}",
+        )
 
 
-def _get_span_name(operation: str, topic: str):
+def _get_span_name(operation: str, topic: str) -> str:
     return f"{topic} {operation}"
 
 
 def _wrap_send(tracer: Tracer) -> Callable:
-    async def _traced_send(func, instance: StreamEngine, args, kwargs):
+    async def _traced_send(func, instance: StreamEngine, args, kwargs) -> Any:
```

Comment:
I think the return type should be `typing.Awaitable[RecordMetadata]`

Reply:
thanks!!